### DISCUSSION TITLE
PubSub message ordering

### DISCUSSION THREAD

**Discussion by WalManas** [processone/ejabberd#3931] [General]
Hello,

I have a question about PubSub message ordering.

I searched stack overflow and I found the [following](https://stackoverflow.com/questions/62628807/how-does-ejabberd-handle-message-ordering-and-delivery#:~:text=First%20of%20all,TCP%20network%20stack.) about ordering in ejabberd. Let us assume that we have a PubSub node with 3 publishers and 5 subscribers, will all of the subscribers observe the messages in the same order? I know that messages of multiple publishers are mixed, but the question is that will all of the subscribers see the same order?

Thanks,

Wael.

**Comment by badlop** [comment]
> Let us assume that we have a PubSub node with 3 publishers and 5 subscribers, will all of the subscribers observe the messages in the same order?

I'd say in most cases probably yes, but in rare circumstances it may not happen.

When a publish query is received by ejabberd, it calls many mod_pubsub functions, including: `publish_item`, which calls `broadcast_publish_item`, which calls `broadcast_stanza`. That last one is the function that loops over the subscribers list and calls `ejabberd_router:route` for each destination.

This is done by the erlang process that handles the publisher session. This means that, if two clients are connected, each one is a publisher, and each one publishes at the same time, each of its erlang processes will run those functions at the same time, in parallel. Probably one of them is processed slightly earlier, so all the destinations will receive their stanzas slightly earlier than the other.

Would you like to have a central process (in the pubsub service or in ejabberd) that manages the XMPP IQ queries, and waits for one to end before accepting the next one?
